src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,368 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Copyright (c) Geoff Norton. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information. 
+//
+
+// ==++==
+//
+
+//
+// ==--==
+
+#include "unixasmmacros.inc"
+#include "asmconstants.h"
+
+//
+// METHODDESC_REGISTER: UMEntryThunk*
+//
+NESTED_ENTRY TheUMEntryPrestub, _TEXT, UMEntryPrestubUnwindFrameChainHandler
+
+#define TheUMEntryPrestub_XMM_SAVE_OFFSET SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES
+
+// Ensure that the new rsp will be 16-byte aligned.  Note that the caller has
+// already pushed the return address.
+#if (((SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES + SIZEOF_MAX_FP_ARG_SPILL) + 8) % 16) != 0
+#define TheUMEntryPrestub_STACK_FRAME_SIZE (SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES + SIZEOF_MAX_FP_ARG_SPILL + 8)
+#else
+#define TheUMEntryPrestub_STACK_FRAME_SIZE (SIZEOF_MAX_OUTGOING_ARGUMENT_HOMES + SIZEOF_MAX_FP_ARG_SPILL)
+#endif
+
+        alloc_stack     TheUMEntryPrestub_STACK_FRAME_SIZE
+
+        save_reg_postrsp    rdi, TheUMEntryPrestub_STACK_FRAME_SIZE + 0x8
```

Comment:
Actually, it would be best to simplify the implementation using the macros to just something like:

NESTED_ENTRY TheUMEntryPrestub, _TEXT, UMEntryPrestubUnwindFrameChainHandler

PUSH_ARGUMENT_REGISTERS
// +8 for alignment
alloc_stack SIZEOF_MAX_FP_ARG_SPILL + 8
SAVE_FLOAT_ARGUMENT_REGISTERS 0
END_PROLOGUE

mov             rdi, METHODDESC_REGISTER 
call            C_FUNC(TheUMEntryPrestubWorker) 

// we're going to tail call to the exec stub that we just setup

RESTORE_FLOAT_ARGUMENT_REGISTERS 0
free_stack SIZEOF_MAX_FP_ARG_SPILL + 8
POP_ARGUMENT_REGISTERS
TAILJMP_RAX